File: module-app/src/main/java/maple/expectation/dto/v4/EquipmentExpectationResponseV4.java
package maple.expectation.dto.v4;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;
import lombok.Builder;
import lombok.Getter;
import lombok.extern.jackson.Jacksonized;

/**
 * V4 장비 기대값 응답 DTO (#240)
 *
 * @deprecated Kotlin으로 마이그레이션되었습니다. {@code
 *     maple.expectation.web.dto.v4.EquipmentExpectationResponseV4}를 사용하세요.
 */
@Deprecated(since = "2026-02-28", forRemoval = true)
@Getter
@Builder
@Jacksonized
public class EquipmentExpectationResponseV4 {

  // ==================== 기본 정보 ====================

  private final String userIgn;
  private final LocalDateTime calculatedAt;
  private final boolean fromCache;

  // ==================== 최대 기대값 프리셋 (#240 V4: 합산 → 최대값) ====================

  private final BigDecimal totalExpectedCost; // 3개 프리셋 중 최대 기대값
  private final String totalCostText; // "12조 3456억 7890만" (#240 V4)
  private final CostBreakdownDto totalCostBreakdown;
  private final int maxPresetNo; // 최대 기대값 프리셋 번호 (1, 2, 3)

  // ==================== 프리셋별 기대값 ====================

  private final List<PresetExpectation> presets;

  // ==================== 내부 DTO ====================

  @Getter
  @Builder
  @Jacksonized
  public static class PresetExpectation {
    private final int presetNo;
    private final BigDecimal totalExpectedCost;
    private final String totalCostText; // "12조 3456억 7890만" (#240 V4)
    private final CostBreakdownDto costBreakdown;
    private final List<ItemExpectationV4> items;
  }

  @Getter
  @Builder
  @Jacksonized
  public static class ItemExpectationV4 {
    private final String itemName;
    private final String itemIcon; // 아이콘 URL (#240 V4)
    private final String itemPart;
    private final int itemLevel;
    private final BigDecimal expectedCost;
    private final String expectedCostText; // "500억" (#240 V4)
    private final CostBreakdownDto costBreakdown;
    private final String enhancePath;

    // 잠재능력 정보
    private final String potentialGrade;
    private final String additionalPotentialGrade;

    // 스타포스 정보
    private final int currentStar;
    private final int targetStar;
    private final boolean isNoljang; // 놀장 여부 (#240 V4)

    // 특수 스킬 반지 레벨 (리스트레인트링, 컨티뉴어스링 등)
    private final int specialRingLevel;

    // 큐브별 기대값 (#240 V4)
    private final CubeExpectationDto blackCubeExpectation;
    private final CubeExpectationDto additionalCubeExpectation;

    // 스타포스 옵션별 기대값 (#240)
    private final StarforceExpectationDto starforceExpectation;

    // 환생의 불꽃 기대값
    private final FlameExpectationDto flameExpectation;
  }

  /**
   * 큐브 기대값 DTO (#240 V4)
   *
   * <p>블랙큐브, 에디셔널큐브 각각의 기대값을 표현합니다.
   */
  @Getter
  @Builder
  @Jacksonized
  public static class CubeExpectationDto {
    private final BigDecimal expectedCost;
    private final String expectedCostText; // "5000억"
    private final BigDecimal expectedTrials; // 기대 시도 횟수
    private final String currentGrade; // 현재 등급 (UNIQUE 등)
    private final String targetGrade; // 목표 등급 (LEGENDARY 등)
    private final String potential; // 현재 잠재능력 텍스트 (#240 V4)

    public static CubeExpectationDto empty() {
      return CubeExpectationDto.builder()
          .expectedCost(BigDecimal.ZERO)
          .expectedCostText("0")
          .expectedTrials(BigDecimal.ZERO)
          .potential("")
          .build();
    }
  }

  /**
   * 스타포스 기대값 DTO (#240)
   *
   * <p>ALL 파괴방지 O / ALL 파괴방지 X 두 가지 케이스 제공
   */
  @Getter
  @Builder
  @Jacksonized
  public static class StarforceExpectationDto {

    // 현재/목표 스타 (#240 V4)
    private final int currentStar;
    private final int targetStar;
    private final boolean isNoljang;

    // 파괴방지 X 케이스 (기본)
    private final BigDecimal costWithoutDestroyPrevention;
    private final String costWithoutDestroyPreventionText; // "500억" (#240 V4)
    private final BigDecimal expectedDestroyCountWithout;

    // 파괴방지 O 케이스 (15-17성 적용)
    private final BigDecimal costWithDestroyPrevention;
    private final String costWithDestroyPreventionText; // "800억" (#240 V4)
    private final BigDecimal expectedDestroyCountWith;

    public static StarforceExpectationDto empty() {
      return StarforceExpectationDto.builder()
          .currentStar(0)
          .targetStar(0)
          .isNoljang(false)
          .costWithoutDestroyPrevention(BigDecimal.ZERO)
          .costWithoutDestroyPreventionText("0")
          .expectedDestroyCountWithout(BigDecimal.ZERO)
          .costWithDestroyPrevention(BigDecimal.ZERO)
          .costWithDestroyPreventionText("0")
          .expectedDestroyCountWith(BigDecimal.ZERO)
          .build();
    }
  }

  /**
   * 환생의 불꽃 기대값 DTO
   *
   * <p>강력한/영원한/심연의 환생의 불꽃 각각의 기대 시도 횟수를 표현합니다.
   */
  @Getter
  @Builder
  @Jacksonized
  public static class FlameExpectationDto {
    private final BigDecimal powerfulFlameTrials; // 강력한 환생의 불꽃 기대 시도 횟수
    private final BigDecimal eternalFlameTrials; // 영원한 환생의 불꽃 기대 시도 횟수
    private final BigDecimal abyssFlameTrials; // 심연의 환생의 불꽃 기대 시도 횟수

    public static FlameExpectationDto empty() {
      return FlameExpectationDto.builder()
          .powerfulFlameTrials(BigDecimal.ZERO)
          .eternalFlameTrials(BigDecimal.ZERO)
          .abyssFlameTrials(BigDecimal.ZERO)
          .build();
    }
  }

  @Getter
  @Builder
  @Jacksonized
  public static class CostBreakdownDto {
    private final BigDecimal blackCubeCost;
    private final BigDecimal redCubeCost;
    private final BigDecimal additionalCubeCost;
    private final BigDecimal starforceCost;

    public static CostBreakdownDto empty() {
      return CostBreakdownDto.builder()
          .blackCubeCost(BigDecimal.ZERO)
          .redCubeCost(BigDecimal.ZERO)
          .additionalCubeCost(BigDecimal.ZERO)
          .starforceCost(BigDecimal.ZERO)
          .build();
    }

    public static CostBreakdownDto from(
        maple.expectation.service.v2.calculator.v4.EquipmentExpectationCalculator.CostBreakdown
            breakdown) {
      return CostBreakdownDto.builder()
          .blackCubeCost(breakdown.blackCubeCost())
          .redCubeCost(breakdown.redCubeCost())
          .additionalCubeCost(breakdown.additionalCubeCost())
          .starforceCost(breakdown.starforceCost())
          .build();
    }

    public CostBreakdownDto add(CostBreakdownDto other) {
      return CostBreakdownDto.builder()
          .blackCubeCost(this.blackCubeCost.add(other.blackCubeCost))
          .redCubeCost(this.redCubeCost.add(other.redCubeCost))
          .additionalCubeCost(this.additionalCubeCost.add(other.additionalCubeCost))
          .starforceCost(this.starforceCost.add(other.starforceCost))
          .build();
    }
  }
}

File: module-web/src/main/kotlin/maple/expectation/web/dto/v4/EquipmentExpectationResponseV4.kt
package maple.expectation.web.dto.v4

import com.fasterxml.jackson.annotation.JsonIgnore
import java.time.LocalDateTime

/**
 * V4 장비 기대값 응답 DTO (#240)
 */
data class EquipmentExpectationResponseV4(
    val userIgn: String,
    val calculatedAt: LocalDateTime,
    val fromCache: Boolean,
    val totalExpectedCost: Double,
    val totalCostText: String,
    val totalCostBreakdown: CostBreakdownDto,
    val maxPresetNo: Int,
    val presets: List<PresetExpectation>,
) {
    /** Java 호환성을 위한 메서드 */
    @JsonIgnore
    fun isFromCache(): Boolean = fromCache

    companion object {
        @JvmStatic
        fun builder() = Builder()

        private fun <T : Any> required(value: T?, fieldName: String): T = requireNotNull(value) { "Missing required field: $fieldName" }
    }

    class Builder {
        private var userIgn: String? = null
        private var calculatedAt: LocalDateTime? = null
        private var fromCache: Boolean = false
        private var totalExpectedCost: Double? = null
        private var totalCostText: String? = null
        private var totalCostBreakdown: CostBreakdownDto? = null
        private var maxPresetNo: Int = 0
        private var presets: List<PresetExpectation>? = null

        fun userIgn(v: String) = apply { userIgn = v }
        fun calculatedAt(v: LocalDateTime) = apply { calculatedAt = v }
        fun fromCache(v: Boolean) = apply { fromCache = v }
        fun totalExpectedCost(v: Double) = apply { totalExpectedCost = v }
        fun totalCostText(v: String) = apply { totalCostText = v }
        fun totalCostBreakdown(v: CostBreakdownDto) = apply { totalCostBreakdown = v }
        fun maxPresetNo(v: Int) = apply { maxPresetNo = v }
        fun presets(v: List<PresetExpectation>) = apply { presets = v }
        fun build() = EquipmentExpectationResponseV4(
            userIgn = required(userIgn, "userIgn"),
            calculatedAt = required(calculatedAt, "calculatedAt"),
            fromCache = fromCache,
            totalExpectedCost = required(totalExpectedCost, "totalExpectedCost"),
            totalCostText = required(totalCostText, "totalCostText"),
            totalCostBreakdown = required(totalCostBreakdown, "totalCostBreakdown"),
            maxPresetNo = maxPresetNo,
            presets = required(presets, "presets"),
        )
    }

    data class PresetExpectation(
        val presetNo: Int,
        val totalExpectedCost: Double,
        val totalCostText: String,
        val costBreakdown: CostBreakdownDto,
        val items: List<ItemExpectationV4>,
    )

    data class ItemExpectationV4(
        val itemName: String,
        val itemIcon: String,
        val itemPart: String,
        val itemLevel: Int,
        val expectedCost: Double,
        val expectedCostText: String,
        val costBreakdown: CostBreakdownDto,
        val enhancePath: String,
        val potentialGrade: String?,
        val additionalPotentialGrade: String?,
        val currentStar: Int,
        val targetStar: Int,
        val isNoljang: Boolean,
        val specialRingLevel: Int,
        val blackCubeExpectation: CubeExpectationDto,
        val additionalCubeExpectation: CubeExpectationDto,
        val starforceExpectation: StarforceExpectationDto,
        val flameExpectation: FlameExpectationDto,
    ) {
        companion object {
            @JvmStatic
            fun builder() = ItemBuilder()
        }

        class ItemBuilder {
            private var itemName: String? = null
            private var itemIcon: String? = null
            private var itemPart: String? = null
            private var itemLevel: Int = 0
            private var expectedCost: Double? = null
            private var expectedCostText: String? = null
            private var costBreakdown: CostBreakdownDto? = null
            private var enhancePath: String? = null
            private var potentialGrade: String? = null
            private var additionalPotentialGrade: String? = null
            private var currentStar: Int = 0
            private var targetStar: Int = 0
            private var isNoljang: Boolean = false
            private var specialRingLevel: Int = 0
            private var blackCubeExpectation: CubeExpectationDto? = null
            private var additionalCubeExpectation: CubeExpectationDto? = null
            private var starforceExpectation: StarforceExpectationDto? = null
            private var flameExpectation: FlameExpectationDto? = null

            fun itemName(v: String) = apply { itemName = v }
            fun itemIcon(v: String) = apply { itemIcon = v }
            fun itemPart(v: String) = apply { itemPart = v }
            fun itemLevel(v: Int) = apply { itemLevel = v }
            fun expectedCost(v: Double) = apply { expectedCost = v }
            fun expectedCostText(v: String) = apply { expectedCostText = v }
            fun costBreakdown(v: CostBreakdownDto) = apply { costBreakdown = v }
            fun enhancePath(v: String) = apply { enhancePath = v }
            fun potentialGrade(v: String?) = apply { potentialGrade = v }
            fun additionalPotentialGrade(v: String?) = apply { additionalPotentialGrade = v }
            fun currentStar(v: Int) = apply { currentStar = v }
            fun targetStar(v: Int) = apply { targetStar = v }
            fun isNoljang(v: Boolean) = apply { isNoljang = v }
            fun specialRingLevel(v: Int) = apply { specialRingLevel = v }
            fun blackCubeExpectation(v: CubeExpectationDto) = apply { blackCubeExpectation = v }
            fun additionalCubeExpectation(v: CubeExpectationDto) = apply { additionalCubeExpectation = v }
            fun starforceExpectation(v: StarforceExpectationDto) = apply { starforceExpectation = v }
            fun flameExpectation(v: FlameExpectationDto) = apply { flameExpectation = v }
            fun build() = ItemExpectationV4(
                itemName = required(itemName, "itemName"),
                itemIcon = required(itemIcon, "itemIcon"),
                itemPart = required(itemPart, "itemPart"),
                itemLevel = itemLevel,
                expectedCost = required(expectedCost, "expectedCost"),
                expectedCostText = required(expectedCostText, "expectedCostText"),
                costBreakdown = required(costBreakdown, "costBreakdown"),
                enhancePath = required(enhancePath, "enhancePath"),
                potentialGrade = potentialGrade,
                additionalPotentialGrade = additionalPotentialGrade,
                currentStar = currentStar,
                targetStar = targetStar,
                isNoljang = isNoljang,
                specialRingLevel = specialRingLevel,
                blackCubeExpectation = required(blackCubeExpectation, "blackCubeExpectation"),
                additionalCubeExpectation = required(additionalCubeExpectation, "additionalCubeExpectation"),
                starforceExpectation = required(starforceExpectation, "starforceExpectation"),
                flameExpectation = required(flameExpectation, "flameExpectation"),
            )
        }
    }

    data class CubeExpectationDto(
        val expectedCost: Double,
        val expectedCostText: String,
        val expectedTrials: Double,
        val currentGrade: String,
        val targetGrade: String,
        val potential: String,
    ) {
        companion object {
            @JvmStatic
            fun empty() = CubeExpectationDto(
                expectedCost = 0.0,
                expectedCostText = "0",
                expectedTrials = 0.0,
                currentGrade = "",
                targetGrade = "",
                potential = "",
            )

            @JvmStatic
            fun builder() = CubeBuilder()
        }

        class CubeBuilder {
            private var expectedCost: Double? = null
            private var expectedCostText: String? = null
            private var expectedTrials: Double? = null
            private var currentGrade: String? = null
            private var targetGrade: String? = null
            private var potential: String? = null

            fun expectedCost(v: Double) = apply { expectedCost = v }
            fun expectedCostText(v: String) = apply { expectedCostText = v }
            fun expectedTrials(v: Double) = apply { expectedTrials = v }
            fun currentGrade(v: String) = apply { currentGrade = v }
            fun targetGrade(v: String) = apply { targetGrade = v }
            fun potential(v: String) = apply { potential = v }
            fun build() = CubeExpectationDto(
                expectedCost = required(expectedCost, "expectedCost"),
                expectedCostText = required(expectedCostText, "expectedCostText"),
                expectedTrials = required(expectedTrials, "expectedTrials"),
                currentGrade = required(currentGrade, "currentGrade"),
                targetGrade = required(targetGrade, "targetGrade"),
                potential = required(potential, "potential"),
            )
        }
    }

    data class StarforceExpectationDto(
        val currentStar: Int,
        val targetStar: Int,
        val isNoljang: Boolean,
        val costWithoutDestroyPrevention: Double,
        val costWithoutDestroyPreventionText: String,
        val expectedDestroyCountWithout: Double,
        val costWithDestroyPrevention: Double,
        val costWithDestroyPreventionText: String,
        val expectedDestroyCountWith: Double,
    ) {
        companion object {
            @JvmStatic
            fun empty() = StarforceExpectationDto(
                currentStar = 0,
                targetStar = 0,
                isNoljang = false,
                costWithoutDestroyPrevention = 0.0,
                costWithoutDestroyPreventionText = "0",
                expectedDestroyCountWithout = 0.0,
                costWithDestroyPrevention = 0.0,
                costWithDestroyPreventionText = "0",
                expectedDestroyCountWith = 0.0,
            )

            @JvmStatic
            fun builder() = StarforceBuilder()
        }

        class StarforceBuilder {
            private var currentStar: Int = 0
            private var targetStar: Int = 0
            private var isNoljang: Boolean = false
            private var costWithoutDestroyPrevention: Double? = null
            private var costWithoutDestroyPreventionText: String? = null
            private var expectedDestroyCountWithout: Double? = null
            private var costWithDestroyPrevention: Double? = null
            private var costWithDestroyPreventionText: String? = null
            private var expectedDestroyCountWith: Double? = null

            fun currentStar(v: Int) = apply { currentStar = v }
            fun targetStar(v: Int) = apply { targetStar = v }
            fun isNoljang(v: Boolean) = apply { isNoljang = v }
            fun costWithoutDestroyPrevention(v: Double) = apply { costWithoutDestroyPrevention = v }
            fun costWithoutDestroyPreventionText(v: String) = apply { costWithoutDestroyPreventionText = v }
            fun expectedDestroyCountWithout(v: Double) = apply { expectedDestroyCountWithout = v }
            fun costWithDestroyPrevention(v: Double) = apply { costWithDestroyPrevention = v }
            fun costWithDestroyPreventionText(v: String) = apply { costWithDestroyPreventionText = v }
            fun expectedDestroyCountWith(v: Double) = apply { expectedDestroyCountWith = v }
            fun build() = StarforceExpectationDto(
                currentStar = currentStar,
                targetStar = targetStar,
                isNoljang = isNoljang,
                costWithoutDestroyPrevention = required(costWithoutDestroyPrevention, "costWithoutDestroyPrevention"),
                costWithoutDestroyPreventionText = required(costWithoutDestroyPreventionText, "costWithoutDestroyPreventionText"),
                expectedDestroyCountWithout = required(expectedDestroyCountWithout, "expectedDestroyCountWithout"),
                costWithDestroyPrevention = required(costWithDestroyPrevention, "costWithDestroyPrevention"),
                costWithDestroyPreventionText = required(costWithDestroyPreventionText, "costWithDestroyPreventionText"),
                expectedDestroyCountWith = required(expectedDestroyCountWith, "expectedDestroyCountWith"),
            )
        }
    }

    data class FlameExpectationDto(
        val powerfulFlameTrials: Double,
        val eternalFlameTrials: Double,
        val abyssFlameTrials: Double,
    ) {
        companion object {
            @JvmStatic
            fun empty() = FlameExpectationDto(
                powerfulFlameTrials = 0.0,
                eternalFlameTrials = 0.0,
                abyssFlameTrials = 0.0,
            )

            @JvmStatic
            fun builder() = FlameBuilder()
        }

        class FlameBuilder {
            private var powerfulFlameTrials: Double? = null
            private var eternalFlameTrials: Double? = null
            private var abyssFlameTrials: Double? = null

            fun powerfulFlameTrials(v: Double) = apply { powerfulFlameTrials = v }
            fun eternalFlameTrials(v: Double) = apply { eternalFlameTrials = v }
            fun abyssFlameTrials(v: Double) = apply { abyssFlameTrials = v }
            fun build() = FlameExpectationDto(
                powerfulFlameTrials = required(powerfulFlameTrials, "powerfulFlameTrials"),
                eternalFlameTrials = required(eternalFlameTrials, "eternalFlameTrials"),
                abyssFlameTrials = required(abyssFlameTrials, "abyssFlameTrials"),
            )
        }
    }

    data class CostBreakdownDto(
        val blackCubeCost: Double,
        val redCubeCost: Double,
        val additionalCubeCost: Double,
        val starforceCost: Double,
    ) {
        companion object {
            @JvmStatic
            fun empty() = CostBreakdownDto(
                blackCubeCost = 0.0,
                redCubeCost = 0.0,
                additionalCubeCost = 0.0,
                starforceCost = 0.0,
            )
        }

        fun add(other: CostBreakdownDto): CostBreakdownDto = copy(
            blackCubeCost = blackCubeCost + other.blackCubeCost,
            redCubeCost = redCubeCost + other.redCubeCost,
            additionalCubeCost = additionalCubeCost + other.additionalCubeCost,
            starforceCost = starforceCost + other.starforceCost,
        )
    }
}
